docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,99 @@
+    ConnectionStateDisconnecting     // Corresponds to controller state:
+                                     // StateDisconnecting
+  };
+```
+
+If at any point any of the checks in `ConnectionManager::ConnectionStates` fails then we will emit a signal to the frontend explaining what exactly the issue is. For example, if we are in `ConnectionManager::ConnectionStateInternetProbe` and the internet probe fails, we will notify the frontend via `emit connectionStateInternetProbeFailed()` signal. This allows the UI to notify the user that the VPN activation failed due to their network connectivity issue and may offer troubleshooting advice such as checking that their wifi/cellular is enabled. It is worth noting that in any given circumstance, we will not deactivate the VPN on our end regardless of the failure. 
+
+This change also means that prior to the activation of the VPN, there are multiple checks and probes that need to succeed; this will take some time (hopefully no more than a couple of seconds), and as long as we communicate to the user via front end prompts about what is happening, this is acceptable.
+
+This is a big undertaking that will take multiple sprints to complete. This document does not cover the extent of work required for creating new unit and functional tests as well as updating the existing ones that will be affected. I propose we break down the work like below:
+
+1. Create a new `ConnectionManager` object with relevant states. Initially, this will have a 1:1 map to the controller so we can debug and ensure the component works as intended. As we add the logic to probe for various scenarios and move further along the implementation work, we can remove them from the `Controller`.
+2. Add probe internet to `ConnectionManager`: This will allow us to check if there is an active internet connection.
+3. Add firewall probe to `ConnectionManager`: To make sure there are no firewall issues blocking access to the internet.
+4. Add captive portal probe to `ConnectionManager`: This will ensure that the user is not stuck behind a captive portal, which can interfere with internet access.
+5. Add check subscription to `ConnectionManager`: This will allow us to verify that the user has an active subscription before proceeding.
+6. Add server probing to `ConnectionManager`: To ensure that the servers within the user's selected location are available and running.
+7. Get rid of the `ConnectionHealth` object
+8. Audit and cleanup `Controller::stateChanged`. We need to audit all areas where code is listening to `Controller::stateChanged` and see if it should be monitoring the `ConnectionManager` instead, if that is the case, make the necessary changes. [Here](https://searchfox.org/mozilla-vpn-client/search?q=Controller%3A%3AstateChanged&path=&case=false&regexp=false) is a current list of every instance we should audit.
+9. Move timestamp code to its own class/object
+
+
+### Audit and cleanup `Controller::stateChanged`
```

Comment:
Auditing is not described in this section. Let's maybe move that part out into a section before this one about monitoring and validation post-release using telemetry and/or Sentry.